# Bangalore Tech Salary Decoder — Live Project
### [Vaibhav Singh] | The Unlox Academy

In [1]:
import pandas as pd
import numpy as np


# Load the raw dataset for initial inspection
df = pd.read_csv(r"C:\Users\isvai\Downloads\bangalore_tech_salaries.csv")
df.head()

,Employee_ID,Role,years_exp,Current_CTC,Previous_CTC,Company,company_TYPE,Skills,Location,Education_Tier,Joining_Year,Work_Mode
0,BLR0065,DS,6,49.4,32.4 LPA,Ola,Unicorn,"JIRA, JavaScript, Spring Boot, Figma, NumPy, A...",HSR Layout,Tier-1,2022,Remote
1,BLR0080,DevOps,0,9.7,NaN,Walmart Labs,MNC,"Tableau, Python, Figma, JavaScript, TensorFlow",Electronic City,2,2024,Remote
2,BLR0166,SDE Backend,6,"3,360,000",24.2 LPA,Postman,Mid-size,"Agile, Excel, GCP, MongoDB, Java, ML",Whitefield,Tier-1,2023,Hybrid
3,BLR0219,SDE Frontend,8,41.4 LPA,30.3,Amazon India,MNC,"PyTorch, Redis, System Design",MG Road,Tier-2,2022,Work from Office
4,BLR0491,Product Lead,5,"3,640,000",30.6 LPA,QuickPay AI,early-stage,"Tableau, Deep Learning, Agile",Whitefield,Tier-2,2023,Onsite


In [2]:
# Inspect structure, types, missing values, and duplicates
df.info()
print(df.shape)
print(df.isnull().sum())
print(df.duplicated().sum())

<class 'pandas.DataFrame'>
RangeIndex: 1015 entries, 0 to 1014
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   Employee_ID     1015 non-null   str  
 1   Role            1015 non-null   str  
 2   years_exp       1015 non-null   int64
 3   Current_CTC     1015 non-null   str  
 4   Previous_CTC    815 non-null    str  
 5   Company         1015 non-null   str  
 6   company_TYPE    1015 non-null   str  
 7   Skills          988 non-null    str  
 8   Location        995 non-null    str  
 9   Education_Tier  1015 non-null   str  
 10  Joining_Year    1015 non-null   int64
 11  Work_Mode       1015 non-null   str  
dtypes: int64(2), str(10)
memory usage: 202.7 KB
(1015, 12)
Employee_ID         0
Role                0
years_exp           0
Current_CTC         0
Previous_CTC      200
Company             0
company_TYPE        0
Skills             27
Location           20
Education_Tier      0
Joining_Year        0


In [3]:
print(df.columns.tolist())


['Employee_ID', 'Role ', 'years_exp', 'Current_CTC', 'Previous_CTC', 'Company', 'company_TYPE', 'Skills', 'Location', 'Education_Tier', 'Joining_Year', 'Work_Mode']


In [4]:
# Strip stray whitespace from column names (e.g. 'Role ' -> 'Role')
df.columns = df.columns.str.strip()
print(df.columns.tolist())

['Employee_ID', 'Role', 'years_exp', 'Current_CTC', 'Previous_CTC', 'Company', 'company_TYPE', 'Skills', 'Location', 'Education_Tier', 'Joining_Year', 'Work_Mode']


In [5]:
print(df['Role'].unique())
print(df['company_TYPE'].unique())
print(df['Education_Tier'].unique())

<ArrowStringArray>
[                       'DS',                    'DevOps',
               'SDE Backend',              'SDE Frontend',
              'Product Lead',           'Product Manager',
 'Site Reliability Engineer',                        'DA',
                     'UI/UX',              'SDE-Frontend',
               'UI Designer',               'ML Engineer',
              'Frontend Dev',            'Data Scientist',
       'Full Stack Engineer',                  'Designer',
          'Product Designer',                        'BA',
         'Frontend Engineer',                'BI Analyst',
               'Backend Dev',                     'Sr PM',
                        'FE',               'SDE-Backend',
              'Data Analyst',          'Business Analyst',
         'Backend Developer',  'Business Systems Analyst',
     'Data Science Engineer',                 'FullStack',
                        'BE',        'Frontend Developer',
          'Backend Engineer',        

In [6]:
# Mapping observed Role variants to canonical role names, based on df['Role'].unique() inspection
# Mapping every observed Role variant to 9 canonical roles, based on df['Role'].unique() inspection
role_map = {
    # SDE Backend
    'SDE Backend': 'SDE Backend',
    'Backend Dev': 'SDE Backend',
    'Backend Developer': 'SDE Backend',
    'Backend Engineer': 'SDE Backend',
    'BE': 'SDE Backend',
    'SDE-Backend': 'SDE Backend',

    # SDE Frontend
    'SDE Frontend': 'SDE Frontend',
    'Frontend Dev': 'SDE Frontend',
    'Frontend Developer': 'SDE Frontend',
    'Frontend Engineer': 'SDE Frontend',
    'FE': 'SDE Frontend',
    'SDE-Frontend': 'SDE Frontend',

    # SDE Full-Stack
    'Full Stack Engineer': 'SDE Full-Stack',
    'FullStack': 'SDE Full-Stack',
    'Fullstack Dev': 'SDE Full-Stack',
    'SDE FS': 'SDE Full-Stack',
    'SDE Full-Stack': 'SDE Full-Stack',

    # Data Scientist
    'DS': 'Data Scientist',
    'Data Scientist': 'Data Scientist',
    'Data Science Engineer': 'Data Scientist',
    'ML Engineer': 'Data Scientist',

    # Data Analyst
    'DA': 'Data Analyst',
    'Data Analyst': 'Data Analyst',
    'Analytics Engineer': 'Data Analyst',

    # Business Analyst
    'BA': 'Business Analyst',
    'Business Analyst': 'Business Analyst',
    'Business Systems Analyst': 'Business Analyst',
    'BI Analyst': 'Business Analyst',

    # Product Manager
    'Product Manager': 'Product Manager',
    'Product Lead': 'Product Manager',
    'PM': 'Product Manager',
    'Sr PM': 'Product Manager',

    # UI/UX Designer
    'UI/UX': 'UI/UX Designer',
    'UI Designer': 'UI/UX Designer',
    'Product Designer': 'UI/UX Designer',
    'Designer': 'UI/UX Designer',
    'UX Designer': 'UI/UX Designer',

    # DevOps Engineer
    'DevOps': 'DevOps Engineer',
    'DevOps Engineer': 'DevOps Engineer',
    'Site Reliability Engineer': 'DevOps Engineer',
    'SRE': 'DevOps Engineer',
    'Infra Engineer': 'DevOps Engineer',
}


    # ... continue for every variant you saw, using the groupings above


df['role_clean'] = df['Role'].replace(role_map)
print(df['role_clean'].value_counts())

role_clean
Business Analyst    140
Data Scientist      128
DevOps Engineer     127
SDE Backend         116
UI/UX Designer      116
Product Manager     113
SDE Frontend        105
SDE Full-Stack       97
Data Analyst         73
Name: count, dtype: int64


In [7]:
print(len(role_map))  # should be 42 if you mapped every variant

42


In [8]:
df['role_clean'] = df['Role'].replace(role_map)
print(df['role_clean'].nunique())  # should now be 9, not 42

9


In [9]:
# Standardise company type casing (e.g. 'early-stage' vs 'Early-stage')
df['company_type_clean'] = df['company_TYPE'].str.strip().str.title()
print(df['company_type_clean'].value_counts())

company_type_clean
Mnc            348
Mid-Size       261
Unicorn        207
Early-Stage    199
Name: count, dtype: int64


In [10]:
company_map = {
    'unicorn': 'Unicorn', 'Unicorn': 'Unicorn',
    'mnc': 'MNC', 'MNC': 'MNC',
    'mid-size': 'Mid-size', 'Mid-size': 'Mid-size',
    'early-stage': 'Early-stage', 'Early-stage': 'Early-stage',
}
df['company_type_clean'] = df['company_TYPE'].str.strip().str.lower().map(
    lambda x: company_map.get(x, x)
)
print(df['company_type_clean'].value_counts())

company_type_clean
MNC            348
Mid-size       261
Unicorn        207
Early-stage    199
Name: count, dtype: int64


In [11]:
print(df['Education_Tier'].unique())

<ArrowStringArray>
['Tier-1',      '2', 'Tier-2', 'Tier 1', 'Tier 2',     'T3', 'Tier-3',
     'T2',      '1',     'T1', 'Tier 3',      '3']
Length: 12, dtype: str


In [12]:
# Standardise Education_Tier: strip stray whitespace, then map variants to 'Tier 1/2/3'
education_map = {
    'Tier-1': 'Tier 1', 'Tier 1': 'Tier 1', 'T1': 'Tier 1', '1': 'Tier 1',
    'Tier-2': 'Tier 2', 'Tier 2': 'Tier 2', 'T2': 'Tier 2', '2': 'Tier 2',
    'Tier-3': 'Tier 3', 'Tier 3': 'Tier 3', 'T3': 'Tier 3', '3': 'Tier 3',
}

df['education_tier_clean'] = df['Education_Tier'].str.strip().replace(education_map)
print(df['education_tier_clean'].value_counts())

education_tier_clean
Tier 2    510
Tier 3    309
Tier 1    196
Name: count, dtype: int64


In [13]:
def parse_ctc(value):
    """
    Parses Current_CTC / Previous_CTC strings into a float in LPA (Lakhs Per Annum).
    Formats seen: '15.5 LPA', '15.5', '15,50,000', '₹15.5 LPA'
    Strategy: strip currency symbols/text, remove commas, convert to float,
    then check magnitude — if it looks like a raw rupee figure (large number),
    divide by 100000 to convert to LPA.
    """
    if pd.isna(value):
        return np.nan  # preserve NaN for freshers (Previous_CTC)

    s = str(value)
    s = s.replace('₹', '').replace('LPA', '').replace(',', '').strip()

    try:
        num = float(s)
    except ValueError:
        return np.nan  # unparseable value, coerce to NaN

    # If the number is large, it's a raw rupee figure, not LPA
    if num > 1000:
        num = num / 100000

    return num

# Apply to both CTC columns
df['current_ctc_lpa'] = df['Current_CTC'].apply(parse_ctc)
df['previous_ctc_lpa'] = df['Previous_CTC'].apply(parse_ctc)

print(df[['Current_CTC', 'current_ctc_lpa', 'Previous_CTC', 'previous_ctc_lpa']].head(10))

  Current_CTC  current_ctc_lpa Previous_CTC  previous_ctc_lpa
0        49.4             49.4     32.4 LPA              32.4
1         9.7              9.7          NaN               NaN
2   3,360,000             33.6     24.2 LPA              24.2
3    41.4 LPA             41.4         30.3              30.3
4   3,640,000             36.4     30.6 LPA              30.6
5        32.1             32.1     22.0 LPA              22.0
6        24.6             24.6     19.3 LPA              19.3
7    21.7 LPA             21.7         17.7              17.7
8        21.0             21.0          NaN               NaN
9    26.9 LPA             26.9         18.9              18.9


In [14]:
# Drop exact duplicate rows found during inspection
before = df.shape[0]
df = df.drop_duplicates()
after = df.shape[0]
print(f"Dropped {before - after} duplicate rows. New shape: {df.shape}")

Dropped 15 duplicate rows. New shape: (1000, 17)


In [15]:
# Q3.1: CTC distribution by role — median, mean, min, max, sorted by median descending
role_ctc_stats = df.groupby('role_clean')['current_ctc_lpa'].agg(['median', 'mean', 'min', 'max'])
role_ctc_stats = role_ctc_stats.sort_values('median', ascending=False)
print(role_ctc_stats)

                  median       mean   min   max
role_clean                                     
Product Manager    31.30  34.131250  10.8  80.1
Data Scientist     24.70  27.579200  10.0  75.6
SDE Full-Stack     22.35  25.409375   8.9  71.7
DevOps Engineer    22.10  23.267716   8.8  60.3
SDE Frontend       21.20  22.193333   6.7  84.4
SDE Backend        21.10  23.284210   7.9  55.1
UI/UX Designer     18.90  20.968142   6.2  63.3
Business Analyst   18.30  20.825000   5.2  52.7
Data Analyst       16.90  18.298611   5.6  43.4


In [16]:
# Q3.2: Experience curve for SDE Backend — bin years_exp into 4 bands, compute median CTC per band
sde_backend = df[df['role_clean'] == 'SDE Backend'].copy()

# Define bin edges: 0-1, 2-3, 4-5, 6+
bins = [-1, 1, 3, 5, 100]
labels = ['0-1', '2-3', '4-5', '6+']
sde_backend['exp_band'] = pd.cut(sde_backend['years_exp'], bins=bins, labels=labels)

exp_median = sde_backend.groupby('exp_band', observed=True)['current_ctc_lpa'].median()
print(exp_median)

# Compute % growth from one band to the next
growth = exp_median.pct_change() * 100
print(growth)

exp_band
0-1    11.65
2-3    20.00
4-5    25.85
6+     40.40
Name: current_ctc_lpa, dtype: float64
exp_band
0-1          NaN
2-3    71.673820
4-5    29.250000
6+     56.286267
Name: current_ctc_lpa, dtype: float64


In [17]:
# Q3.3: Skill premium for SDEs — compare median CTC with vs without each skill
sde_roles = ['SDE Backend', 'SDE Frontend', 'SDE Full-Stack']
sdes = df[df['role_clean'].isin(sde_roles)].copy()

skills_to_check = ['AWS', 'ML', 'System Design', 'Kubernetes']

for skill in skills_to_check:
    # .str.contains checks if the skill appears anywhere in the comma-separated Skills string
    # na=False treats NaN skill entries as "does not have the skill"
    has_skill = sdes['Skills'].str.contains(skill, case=False, na=False)
    
    median_with = sdes.loc[has_skill, 'current_ctc_lpa'].median()
    median_without = sdes.loc[~has_skill, 'current_ctc_lpa'].median()
    premium_pct = ((median_with - median_without) / median_without) * 100
    
    print(f"{skill:<15} With: {median_with:.1f} LPA   Without: {median_without:.1f} LPA   Premium: {premium_pct:+.1f}%")

AWS             With: 20.9 LPA   Without: 21.9 LPA   Premium: -4.3%
ML              With: 23.9 LPA   Without: 21.3 LPA   Premium: +12.4%
System Design   With: 25.3 LPA   Without: 21.0 LPA   Premium: +20.5%
Kubernetes      With: 23.2 LPA   Without: 21.5 LPA   Premium: +7.9%


In [18]:
# Q3.4: Company-type premium for SDE Backend — median CTC by company type
target_role = df[df['role_clean'] == 'SDE Backend']

company_premium = target_role.groupby('company_type_clean')['current_ctc_lpa'].median()
company_premium = company_premium.sort_values(ascending=False)
print(company_premium)

# % premium of Unicorn over MNC, same role
unicorn_median = company_premium['Unicorn']
mnc_median = company_premium['MNC']
premium_pct = ((unicorn_median - mnc_median) / mnc_median) * 100
print(f"Unicorn premium over MNC: {premium_pct:+.1f}%")

company_type_clean
Unicorn        27.35
MNC            20.45
Mid-size       19.50
Early-stage    18.60
Name: current_ctc_lpa, dtype: float64
Unicorn premium over MNC: +33.7%


In [19]:
# Q3.5 Step 1: create exp_band for the full dataset (needed to group by role + company_type + exp_band)
bins = [-1, 1, 3, 5, 100]
labels = ['0-1', '2-3', '4-5', '6+']
df['exp_band'] = pd.cut(df['years_exp'], bins=bins, labels=labels)

df['group_median'] = df.groupby(
    ['role_clean', 'company_type_clean', 'exp_band'], observed=True
)['current_ctc_lpa'].transform('median')


df['group_size'] = df.groupby(
    ['role_clean', 'company_type_clean', 'exp_band'], observed=True
)['current_ctc_lpa'].transform('size')


In [20]:
# Step 3: gap = individual's CTC minus their group's median. Negative gap = underpaid relative to peers.
df['gap'] = df['current_ctc_lpa'] - df['group_median']

# Only trust groups with at least 10 members (per brief instructions)
reliable = df[df['group_size'] >= 10]

underpaid = reliable.sort_values('gap').head(10)
print(underpaid[['Employee_ID',
                 'role_clean',
                 'company_type_clean',
                 'years_exp',
                 'current_ctc_lpa',
                 'group_median',
                 'gap']])

    Employee_ID        role_clean company_type_clean  years_exp  \
349     BLR0954  Business Analyst                MNC          6   
810     BLR0615    Data Scientist           Mid-size          6   
460     BLR0141    Data Scientist           Mid-size          6   
904     BLR0925       SDE Backend                MNC          4   
250     BLR0072       SDE Backend                MNC          4   
402     BLR0502   DevOps Engineer            Unicorn          2   
420     BLR0768   DevOps Engineer            Unicorn          2   
728     BLR0301   Product Manager                MNC          2   
237     BLR0198   DevOps Engineer                MNC          2   
249     BLR0733  Business Analyst                MNC          7   

     current_ctc_lpa  group_median    gap  
349             25.4         37.65 -12.25  
810             28.1         40.30 -12.20  
460             31.7         40.30  -8.60  
904             19.4         27.30  -7.90  
250             19.5         27.30  -7.80 

In [21]:
print(df.columns.tolist())

['Employee_ID', 'Role', 'years_exp', 'Current_CTC', 'Previous_CTC', 'Company', 'company_TYPE', 'Skills', 'Location', 'Education_Tier', 'Joining_Year', 'Work_Mode', 'role_clean', 'company_type_clean', 'education_tier_clean', 'current_ctc_lpa', 'previous_ctc_lpa', 'exp_band', 'group_median', 'group_size', 'gap']


In [22]:
# Task 4: Final printed report
print("=" * 60)
print(" BANGALORE TECH SALARY DECODER")
print(" Built by Vaibhav Singh | The Unlox Academy | 2-Hour Live Project")
print("=" * 60)
print(f" Dataset : {df.shape[0]} Bengaluru tech professionals (synthetic)")
print(" Period  : 2024 employment snapshot")
print()
print(" ---- MEDIAN CTC BY ROLE (in LPA) ----")
for role, row in role_ctc_stats.iterrows():
    print(f" {role:<20} {row['median']:>6.1f}")
print()
print(" ---- SDE BACKEND CTC BY EXPERIENCE BAND ----")
for band, median in exp_median.items():
    g = growth.get(band, float('nan'))
    growth_str = f" (+{g:.0f}% growth)" if pd.notna(g) else ""
    print(f" {band:<10}: {median:.1f} LPA median{growth_str}")
print()
print(" ---- SKILL PREMIUM FOR SDEs (median CTC) ----")
# (reuse your Q3.3 loop here, printing skill / with / without / premium)
print()
print(" ---- COMPANY-TYPE PREMIUM (SDE Backend, same role) ----")
for ctype, median in company_premium.items():
    print(f" {ctype:<12} {median:.1f} LPA")
print()
print(" ---- TOP 5 MOST-UNDERPAID PROFESSIONALS ----")
for _, r in underpaid.head(5).iterrows():
    print(f" {r['Employee_ID']:<10} {r['role_clean']:<15} {r['company_type_clean']:<10} "
          f"{r['years_exp']} yrs   gap: {r['gap']:+.1f} LPA")
print("=" * 60)

 BANGALORE TECH SALARY DECODER
 Built by Vaibhav Singh | The Unlox Academy | 2-Hour Live Project
 Dataset : 1000 Bengaluru tech professionals (synthetic)
 Period  : 2024 employment snapshot

 ---- MEDIAN CTC BY ROLE (in LPA) ----
 Product Manager        31.3
 Data Scientist         24.7
 SDE Full-Stack         22.4
 DevOps Engineer        22.1
 SDE Frontend           21.2
 SDE Backend            21.1
 UI/UX Designer         18.9
 Business Analyst       18.3
 Data Analyst           16.9

 ---- SDE BACKEND CTC BY EXPERIENCE BAND ----
 0-1       : 11.6 LPA median
 2-3       : 20.0 LPA median (+72% growth)
 4-5       : 25.9 LPA median (+29% growth)
 6+        : 40.4 LPA median (+56% growth)

 ---- SKILL PREMIUM FOR SDEs (median CTC) ----

 ---- COMPANY-TYPE PREMIUM (SDE Backend, same role) ----
 Unicorn      27.4 LPA
 MNC          20.5 LPA
 Mid-size     19.5 LPA
 Early-stage  18.6 LPA

 ---- TOP 5 MOST-UNDERPAID PROFESSIONALS ----
 BLR0954    Business Analyst MNC        6 yrs   gap: -12.2 

## Key Insights

1. Product Managers earn the highest median CTC at 31.3 LPA, 1.85x the median of Data Analysts, 
   the lowest-paid role at 16.9 LPA. Students choosing between PM-track and analyst-track roles 
   should factor in this near-2x gap when negotiating offers or planning a career pivot.

2. SDE Backend CTC jumps +72% between the 0-1 and 2-3 year experience bands (11.6 LPA to 
   20.0 LPA) — the steepest single jump in the career curve. Students should prioritize crossing 
   the 2-year mark before switching jobs, since that's where the market rewards experience most.

3. System Design carries the highest premium among SDEs at +20.5% (25.3 LPA with vs 21.0 LPA 
   without) — notably higher than Kubernetes (+7.9%) or ML (+12.4%), and AWS actually shows a 
   slight negative premium (-4.3%), suggesting it's become table-stakes rather than a 
   differentiator. Students prepping for SDE interviews should prioritize System Design depth 
   over checkbox cloud certifications, since it shows the clearest measurable payoff.